browser-sign-in/__tests__/App-test.js

```diff
@@ -1,12 +1,205 @@
+            user: [{ foo: 'foo', bar: 'bar' }],
+          }
+        },
+        ok: true,
+      });
+    });
+    return promise;
+  })
+  .mockImplementationOnce(() => {
+    const promise = new Promise((resolve, reject) => {
+      resolve({
+        json: () => {
+          return {
+            userinfo_endpoint: 'dummy_endpoint',
+          }
+        },
+        ok: true,
+      });
+    });
+    return promise;
+  });
+
+describe('app setup', () => {
+  it('should render without crashing', () => {
+    const rendered = renderer.create(<App />).toJSON();
+    expect.assertions(1);
+    expect(rendered).toBeTruthy();
+  });
+
+  it('should render correctly', () => {
+    const rendered = renderer.create(<App />).toJSON();
+    expect.assertions(1);
+    expect(rendered).toMatchSnapshot();
+  });
+
+  it('should initialize with default state', () => {
+    const wrapper = shallow(<App />);
+    expect.assertions(2);
+    expect(wrapper.state().authenticated).toBe(false);
+    expect(wrapper.state().context).toBe(null);
+  });
+
+  it('should render login button if not authenticated', () => {
+    const wrapper = shallow(<App />);
+    const loginButton = wrapper.find('Button').get(0);
+    expect.assertions(1);
+    expect(loginButton.props.title).toBe('Login');
+  });
+
+  it('should render logout and get user info buttons if authenticated', () => {
+    const wrapper = shallow(<App />);
+    wrapper.setState({authenticated: true});
+    const logoutButton = wrapper.find('Button').get(0);
+    const getUserFromIdButton = wrapper.find('Button').get(1);
+    const getUserButton = wrapper.find('Button').get(2);
+    const getUserFromTokenButton = wrapper.find('Button').get(3);
+    expect.assertions(4);
+    expect(logoutButton.props.title).toBe('Logout');
+    expect(getUserFromIdButton.props.title).toBe('Get User From Id Token');
+    expect(getUserButton.props.title).toBe('Get User From Request');
+    expect(getUserFromTokenButton.props.title).toBe('Get User From Access Token');
+  });
+
+  it('should not render login button if authenticated', () => {
+    const wrapper = shallow(<App />);
+    wrapper.setState({authenticated: true});
+    const loginButton = wrapper.find('Button').get(0);
+    expect.assertions(1);
+    expect(loginButton.props.title).not.toBe('Login');
+  });
+
+  it('should not render logout and get user info buttons if not authenticated', () => {
+    const wrapper = shallow(<App />);
+    const logoutButton = wrapper.find('Button').get(0);
+    const getUserFromIdButton = wrapper.find('Button').get(1);
+    const getUserButton = wrapper.find('Button').get(2);
+    const getUserFromTokenButton = wrapper.find('Button').get(3);
+    expect.assertions(4);
+    expect(logoutButton.props.title).not.toBe('Logout');
+    expect(getUserFromIdButton).toBe(undefined);
+    expect(getUserButton).toBe(undefined);
+    expect(getUserFromTokenButton).toBe(undefined);
+  });
+});
+
+describe('when authenticated', () => {
```

Comment:
nit: While the english flows fine here, when you actually run the tests (depending on reporter syntax) it will output:

"when authenticated should detect when the user has logged in"
"when authenticated should detect when the user has logged out"
(etc)

I recommend tweaking the describe() string to make those read a bit more clearly.